diff --git a/src/rest/CMakeLists.txt b/src/rest/CMakeLists.txt
index 517a5ea4..e4f9f0b1 100644
--- a/src/rest/CMakeLists.txt
+++ b/src/rest/CMakeLists.txt
@@ -69,5 +69,5 @@ target_link_libraries(otbr-rest
         otbr-utils
         openthread-ftd
         openthread-posix
-        pthread
+        $<$<NOT:$<BOOL:${ANDROID}>>:pthread>
 )
diff --git a/src/rest/json.cpp b/src/rest/json.cpp
index ed5a36d7..6659440f 100644
--- a/src/rest/json.cpp
+++ b/src/rest/json.cpp
@@ -143,7 +143,8 @@ static cJSON *MeshChildrenIp62Json(const std::vector<DeviceIp6Addrs> &aChildrenI
 
 static cJSON *Bytes2HexJson(const uint8_t *aBytes, uint8_t aLength)
 {
-    char hex[2 * aLength + 1];
+    std::vector<char> hexBuf(2 * aLength + 1);
+    char             *hex = hexBuf.data();
 
     otbr::Utils::Bytes2Hex(aBytes, aLength, hex);
     hex[2 * aLength] = '\0';
